runtime/devnet/src/extensions/mod.rs

```diff
@@ -0,0 +1,1160 @@
+	let result = match key {
+		RuntimeStateKeys::Nfts(key) => read_nfts_state::<T, E>(key, env),
+		RuntimeStateKeys::ParachainSystem(key) => read_parachain_system_state::<T, E>(key, env),
+		RuntimeStateKeys::Assets(key) => read_assets_state::<T, E>(key, env),
+	}?
+	.encode();
+	log::trace!(
+		target:LOG_TARGET,
+		"{} result: {:?}.", LOG_PREFIX, result
+	);
+	env.write(&result, false, None)
+}
+
+fn send_xcm<T, E>(env: &mut Environment<E, BufInBufOutState>) -> Result<(), DispatchError>
+where
+	T: pallet_contracts::Config
+		+ frame_system::Config<
+			RuntimeOrigin = RuntimeOrigin,
+			AccountId = AccountId,
+			RuntimeCall = RuntimeCall,
+		>,
+	E: Ext<T = T>,
+{
+	const LOG_PREFIX: &str = " send_xcm |";
+	// Read the input as CrossChainMessage.
+	let xc_call: CrossChainMessage = env.read_as::<CrossChainMessage>()?;
+	// Determine the call to dispatch.
+	let (dest, message) = match xc_call {
+		CrossChainMessage::Relay(message) => {
+			let dest = Location::parent().into_versioned();
+			let assets: Asset = (Here, 10 * UNIT).into();
+			let beneficiary: Location =
+				AccountId32 { id: (env.ext().address().clone()).into(), network: None }.into();
+			let message = Xcm::builder()
+				.withdraw_asset(assets.clone().into())
+				.buy_execution(assets.clone(), Unlimited)
+				.transact(
+					SovereignAccount,
+					Weight::from_parts(250_000_000, 10_000),
+					message.encode().into(),
+				)
+				.refund_surplus()
+				.deposit_asset(assets.into(), beneficiary)
+				.build();
+			(dest, message)
+		},
+	};
+	// TODO: revisit to replace with signed contract origin
+	let origin: RuntimeOrigin = RawOrigin::Root.into();
+	// Generate runtime call to dispatch.
+	let call = RuntimeCall::PolkadotXcm(pallet_xcm::Call::send {
+		dest: Box::new(dest),
+		message: Box::new(VersionedXcm::V4(message)),
+	});
+	dispatch_call::<T, E>(env, call, origin, LOG_PREFIX)
+}
+
+// Converts a `DispatchError` to a `u32` status code based on the version of the API the contract uses.
+// The contract calling the chain extension can convert the status code to the descriptive `Error`.
+//
+// For `Error` see `pop_primitives::<version>::error::Error`.
+//
+// The error encoding can vary per version, allowing for flexible and backward-compatible error handling.
+// As a result, contracts maintain compatibility across different versions of the runtime.
+//
+// # Parameters
+//
+// - `error`: The `DispatchError` encountered during contract execution.
+// - `version`: The version of the chain extension, used to determine the known errors.
+pub(crate) fn convert_to_status_code(error: DispatchError, version: u8) -> u32 {
+	// "UnknownFunctionId" and "DecodingFailed" are mapped to specific errors in the API and will
+	// never change.
+	let mut encoded_error = match error {
+		DispatchError::Other("UnknownFunctionId") => Vec::from([254u8, 0, 0, 0]),
+		DispatchError::Other("DecodingFailed") => Vec::from([255u8, 0, 0, 0]),
```

Comment:
Assume this is the 255 magic number in the pop_api crate. Please create some const there defined as such and update all 255 references to use the const so it is clear what is being emitted as the status code.

Might as well do the same for 254. Its a pity that those are probably already defined on the pop api error enum, but oh well.